elfpy/utils/logs.py

```diff
@@ -23,81 +23,79 @@ def setup_logging(
-    if log_format_string is None:
-        log_formatter = logging.Formatter(elfpy.DEFAULT_LOG_FORMATTER, elfpy.DEFAULT_LOG_DATETIME)
-    else:
-        log_formatter = logging.Formatter(log_format_string, elfpy.DEFAULT_LOG_DATETIME)
-    # create log handlers
-    handlers = logging.getLogger().handlers if keep_previous_handlers else []
-    # pipe to stdout if requested
-    if log_stdout:
-        stream_handler = logging.StreamHandler(sys.stdout)
-        stream_handler.setFormatter(log_formatter)
-        handlers.append(stream_handler)
-    # log to file
+    # pylint: disable=too-many-arguments
+    # remove all handlers if requested
+    if not keep_previous_handlers:
+        _remove_handlers(get_root_logger())
+    # add handler logging to stdout if requested
+    if log_stdout is True:
+        add_stdout_handler(log_format_string=log_format_string, log_level=log_level)
+    # add handler logging to file if requested
     if log_filename is not None:
-        log_dir, log_name = _prepare_log_path(log_filename)
-        # Delete the log file if requested
-        if delete_previous_logs and os.path.exists(os.path.join(log_dir, log_name)):
-            os.remove(os.path.join(log_dir, log_name))
-        file_handler = _create_file_handler(log_dir, log_name, log_formatter, max_bytes)
-        handlers.append(file_handler)
-    # Configure the root logger
-    logger = logging.getLogger()
-    logger.setLevel(log_level)
-    logger.handlers = handlers
+        add_file_handler(
+            log_filename=log_filename,
+            delete_previous_logs=delete_previous_logs,
+            log_format_string=log_format_string,
+            max_bytes=max_bytes,
+            log_level=log_level,
+        )
+    # Set the root logger's level to the lowest level among all of its
+    # While the root logger doesn't log anything, it captures logging statements to feed to the handlers
+    # Therefore set the root logger's level to the lowest level among all of its handlers
+    if get_root_logger().handlers:
+        get_root_logger().setLevel(min(handler.level for handler in get_root_logger().handlers))
+    else:
+        get_root_logger().setLevel(_create_log_level(log_level))
 
 
 def close_logging(delete_logs=True):
-    r"""Close logging and handlers for the test"""
+    r"""Close logging and remove handlers for the test."""
     logging.shutdown()
+    root_logger = get_root_logger()
     if delete_logs:
-        for handler in logging.getLogger().handlers:
+        for handler in root_logger.handlers:
             if hasattr(handler, "baseFilename") and not isinstance(handler, logging.StreamHandler):
                 # access baseFilename in a type safe way
                 handler_file_name = getattr(handler, "baseFilename", None)
                 if handler_file_name is not None and os.path.exists(handler_file_name):
                     os.remove(handler_file_name)
             handler.close()
-    # logging.getLogger().handlers = []
```

Comment:
why did we have this commented out before? did we want to not remove handlers on a `close_logging` call? My assumption here is that after this call, we'll be re-initializing logging from scratch, so we don't want to save anything. That's how the testing works.

Reply:
double-checked that `close` does not remove a handler. but I believe removing it is what we want.